### PR TITLE
Automate the process of creating a Force review app

### PR DESCRIPTION
**Pull request by dleve123**
Automate the process of creating a Force review app
    
    This is definitely a rough cut, but figured something is better than
    nothing. Tried to heavily document so that the script so that when
    things break, at least devs know the intent here.
    
    * Add new build_review_app.sh script to scripts/

### CODE REVIEW

File: docs/review_app_setup.sh [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+# Description: This is a script to help auctomate the process of building
+# Hokusai review application for Force. It draws upon Artsy's existing review
+# app documentation:
+# https://github.com/artsy/hokusai/blob/master/docs/Review_Apps.md
+# and experiences trying to set up review apps!
+
+# Bail out of script on first expression failure.
+set -e
+
+NAME=auct-281
```
Comment (dleve123) [inline docs/review_app_setup.sh]: this should be shell args

File: docs/review_app_setup.sh [hunk 2]
```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+
+# Description: This is a script to help auctomate the process of building
```
Comment (joeyAghion) [inline docs/review_app_setup.sh]: au~c~tomate

File: docs/review_app_setup.sh [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "..."
+echo "yeah, I'm takling to you"
```
Comment (joeyAghion) [inline docs/review_app_setup.sh]: `talking`

File: hokusai/auct-281.yml [hunk 4]
```diff
@@ -0,0 +1,149 @@
+---
```
Comment (joeyAghion) [inline hokusai/auct-281.yml]: Did you mean to commit this file?

Comment (dleve123) [reply hokusai/auct-281.yml]: Want to delete it before merging the PR, just wanted to show reviews an example of a fulled processed review app.

**Comment by dleve123** [comment]
QA:

TL;DR

- took ~11 mins to automate the review app creation process
- ran without warning.

As just @joeyAghion, I'll do a couple of tweaks for dev UX, but I think this is actually a well-documented script, and not a runbook.

```sh
dl@phoenix:~/src/force * (review-app-setup-for-force)
$ time ./docs/review_app_setup.sh auct-281-automated

NAME="$1"

echo $NAME
auct-281-automated

# Generate the Kubernetes YAML needed to provision the application.
hokusai review_app setup $NAME
Created hokusai/auct-281-automated.yml

# Create the Docker image of your current working direct of Force, and push
# it to Artsy's docker registry.
#
# --force is needed as the current working directory is dirty with (at least)
# the YAML file generated above.
# --skip-latest as we're making no claim that this is the "latest" build of the
# service.
# --tag to name the image.
#
echo "WARNING: This is likely going to take ~10 mins on your MBP."
WARNING: This is likely going to take ~10 mins on your MBP.
echo "Be patient and grab some baby carrots."
Be patient and grab some baby carrots.
hokusai registry push --force --skip-latest --verbose --tag $NAME
==> hokusai exec `docker login -u ****** -p ***** https://585031190124.dkr.ecr.us-east-1.amazonaws.com`

==> hokusai exec `docker-compose -f /Users/dl/src/force/hokusai/build.yml -p hokusai build`

Building force
Step 1/11 : FROM node:10.13-alpine
 ---> 93f2dcbcddfe
Step 2/11 : WORKDIR /app
 ---> Using cache
 ---> baba70a7dbb8
Step 3/11 : RUN apk --no-cache --quiet add       bash       build-base       curl       dumb-init       git       python &&       adduser -D -g '' deploy
 ---> Using cache
 ---> ab4b43ecb0ad
Step 4/11 : COPY package.json yarn.lock ./
 ---> Using cache
 ---> 5e842a765257
Step 5/11 : COPY patches ./patches
 ---> Using cache
 ---> 17e07c2bb132
Step 6/11 : RUN yarn install --frozen-lockfile --quiet
 ---> Using cache
 ---> b185f2ec21d5
Step 7/11 : COPY . ./
 ---> a03a910ef451
Step 8/11 : RUN yarn assets &&     yarn build:server &&     chown -R deploy:deploy ./
 ---> Running in 6f812c66d51a
yarn run v1.10.1
$ NODE_ENV=production yarn build
$ yarn clean && yarn webpack
$ rm -rf .cache && rm -f manifest.json && rm -rf public/assets && mkdir -p public/assets && echo '[Force] Cleaned build directory'
[Force] Cleaned build directory
$ node --max_old_space_size=2048 -r dotenv/config -r @babel/register node_modules/.bin/webpack --config ./webpack
Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade caniuse-lite browserslist`

[!] No existing `.cache` directory detected, initial launch will take a while.

Webpack: Starting ...
[Force compileCSS] Compiling...
[Force compileCSS] Complete.

[Force] NODE_ENV=production

[Force] Building client-side production code...
Done in 390.13s.
yarn run v1.10.1
$ BUILD_SERVER=true NODE_ENV=production yarn webpack
$ node --max_old_space_size=2048 -r dotenv/config -r @babel/register node_modules/.bin/webpack --config ./webpack
Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade caniuse-lite browserslist`
Webpack: Starting ...

[Force] NODE_ENV=production

[Force] Building server-side code...
Hash: 026b0af43c9e2197449a
Version: webpack 4.30.0
Time: 30917ms
Built at: 08/15/2019 10:44:07 PM
             Asset      Size  Chunks             Chunk Names
    server.dist.js   766 KiB       0  [emitted]  main
server.dist.js.map  1.69 MiB       0  [emitted]  main
Entrypoint main = server.dist.js server.dist.js.map
  [4] external "express" 42 bytes {0} [built]
 [18] external "regenerator-runtime/runtime" 42 bytes {0} [built]
 [34] external "path" 42 bytes {0} [built]
 [72] external "chalk" 42 bytes {0} [built]
 [73] external "artsy-xapp" 42 bytes {0} [built]
[100] ./src/lib/cache.coffee 2.37 KiB {0} [built]
[180] external "core-js/modules/es6.number.constructor" 42 bytes {0} [built]
[267] ./src/index.js 3.07 KiB {0} [built]
[268] (webpack)/buildin/module.js 497 bytes {0} [built]
[269] external "source-map-support" 42 bytes {0} [built]
[270] external "coffeescript/register" 42 bytes {0} [built]
[271] external "@babel/register" 42 bytes {0} [built]
[272] ./src/lib/datadog.ts 557 bytes {0} [built]
[274] external "require-control" 42 bytes {0} [built]
[275] external "bluebird" 42 bytes {0} [built]
    + 859 hidden modules

WARNING in ./src/lib/setup.js 251:47-54
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./src/index.js
Done in 33.87s.
Removing intermediate container 6f812c66d51a
 ---> 784fb29f94f0
Step 9/11 : USER deploy
 ---> Running in d861785cd173
Removing intermediate container d861785cd173
 ---> f3a3519da29e
Step 10/11 : ENTRYPOINT ["/usr/bin/dumb-init", "--"]
 ---> Running in 92ee28ead4f2
Removing intermediate container 92ee28ead4f2
 ---> d8048cab4736
Step 11/11 : CMD ["yarn", "start"]
 ---> Running in 311dd2f6d59d
Removing intermediate container 311dd2f6d59d
 ---> 1f9d85663c78
Successfully built 1f9d85663c78
Successfully tagged hokusai_force:latest
==> hokusai exec `docker tag hokusai_force:latest 585031190124.dkr.ecr.us-east-1.amazonaws.com/force:auct-281-automated`

==> hokusai exec `docker push 585031190124.dkr.ecr.us-east-1.amazonaws.com/force:auct-281-automated`

The push refers to repository [585031190124.dkr.ecr.us-east-1.amazonaws.com/force]
39a9a85cf40a: Pushed
a7ef794a4700: Pushed
4542ca3b11d1: Layer already exists
1738bfe79d08: Layer already exists
b09aebae497a: Layer already exists
164064f9fa87: Layer already exists
8d0b8ee659aa: Layer already exists
2faeaaebb113: Layer already exists
387bc77dd3f2: Layer already exists
df64d3292fd6: Layer already exists
auct-281-automated: digest: sha256:1bc8ed4b10eae8f2b55994e0a508909d99851e3e1df8cdaca020aaa9e904693d size: 2425
Pushed hokusai_force:latest to 585031190124.dkr.ecr.us-east-1.amazonaws.com/force:auct-281-automated


# Edit the K8S YAML to reference the proper Docker image
# sed -i '' 's/:staging/:auct-281/g' hokusai/auct-281.yml
#
# The weird `-i ''` is needed on MacOS:
# https://stackoverflow.com/a/44864004
# WARNING: This might not work on Linux.
sed -i '' "s/:staging/:$NAME/g" hokusai/$NAME.yml

# Edit the K8S YAML to remove the instructions that enforce that the service
# can only be accessible via Cloudflare
#
# First, remove the `loadBalancerSourceRanges:` line
sed -i '' '/loadBalancer/d' hokusai/$NAME.yml

# Then, delete all the IP address lines.
#
# WARNING: This is a bit sketchy as this will
# delete any line of the form "- [number][number][number].". This is my best
# approx for an regex for an IP address, and I'm sure that there are better
# ones.
sed -i '' "/- [[:digit:]][[:digit:]][[:digit:]]./d" hokusai/$NAME.yml

# Provision the review app
hokusai review_app create $NAME --verbose
namespace "auct-281-automated" created
deployment.extensions "force-web" created
horizontalpodautoscaler.autoscaling "force-web" created
service "force-web" created
Created Kubernetes environment /Users/dl/src/force/hokusai/auct-281-automated.yml

# Copy Force staging's ConfigMap to your review app
hokusai review_app env copy $NAME --verbose
==> hokusai exec `kubectl --context staging --namespace default get configmap force-environment -o yaml`

==> hokusai exec `kubectl --context staging --namespace auct-281-automated apply -f /var/folders/c1/5m0hqhyn16qgq4lylv00vgv80000gq/T/tmptPxeRl`


# Copy the staging-shared nginx config to your review app
hokusai review_app env copy $NAME --configmap nginx-config --verbose
==> hokusai exec `kubectl --context staging --namespace default get configmap nginx-config -o yaml`

==> hokusai exec `kubectl --context staging --namespace auct-281-automated apply -f /var/folders/c1/5m0hqhyn16qgq4lylv00vgv80000gq/T/tmpalxyRq`


# To enable authentication via Force's server, we need to allow XHR requests
# from Force's client to server. As such, Force's server needs to have the
# proper name of the domain that the requests are coming from. Otherwise,
# authentication requests won't work!
hokusai review_app env set $NAME \
  APP_URL="https://$NAME.artsy.net" \
  COOKIE_DOMAIN="$NAME.artsy.net" \
  FORCE_URL="https://$NAME.artsy.net"

# Now you need to create a CNAME for your review app and wait. This is required
# as Gravity only allows authentication requests from requests of originating
# from the artsy.net domain
echo "Hey you human!"
Hey you human!
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "..."
...
echo "yeah, I'm talking to you"
yeah, I'm talking to you
echo "you need to configure a CNAME record so that $NAME.artsy.net resolves to" \
     "the service's external IP:"
you need to configure a CNAME record so that auct-281-automated.artsy.net resolves to the service's external IP:
echo $(kubectl get service force-web --namespace $NAME --context staging -o json \
  | jq .status.loadBalancer.ingress[].hostname)
kubectl get service force-web --namespace $NAME --context staging -o json   | jq .status.loadBalancer.ingress[].hostname
"ac157ee03bfae11e9aff2127c3b176f7-175025084.us-east-1.elb.amazonaws.com"
echo "you may do this in the DynDNS interface. Credentials are in 1pass"
you may do this in the DynDNS interface. Credentials are in 1pass

echo "After that, fini. Thank you for review app-ing"
After that, fini. Thank you for review app-ing

exit 0

real    11m40.583s
user    0m9.388s
sys     0m5.530s
```

**Comment by dleve123** [comment]
This is ready for review.

File: docs/review_app_setup.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+
+# Description: This is a script to help automate the process of building
```
Comment (eessex) [inline docs/review_app_setup.sh]: I think this can be more declarative, "Run this script to automate...". I'd also add a comment with the command to run it, to make this accessible as possible. I'm assuming a user will have to pass the name variable.

Comment (izakp) [reply docs/review_app_setup.sh]: +1 - would document the invocation with name as required argument

File: docs/review_app_setup.sh [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+# https://github.com/artsy/hokusai/blob/master/docs/Review_Apps.md
+# and experiences trying to set up review apps!
+
+echo "[setup_review_app.sh] START"
+
+# Bail out of script on first expression failure and sprint the commands as
+# they are being run.
+set -ev
+
+NAME="$1"
```
Comment (izakp) [inline docs/review_app_setup.sh]: I would bail here if $1 is not provided / `NAME` is empty

File: docs/review_app_setup.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+sed -i '' "/- [[:digit:]][[:digit:]][[:digit:]]./d" hokusai/$NAME.yml
+
+# Provision the review app
+hokusai review_app create $NAME --verbose
+
+# Copy Force staging's ConfigMap to your review app
+hokusai review_app env copy $NAME --verbose
+
+# Copy the staging-shared nginx config to your review app
+hokusai review_app env copy $NAME --configmap nginx-config --verbose
```
Comment (izakp) [inline docs/review_app_setup.sh]: Ah we may need to also copy over Nginx secrets here until we can build in a Hokusai workaround...

```
kubectl --context staging get secret nginx-secrets -o yaml > nginx-secrets.yml
sed -i '' "s/namespace: default/namespace: $NAME/g" nginx-secrets.yml
kubectl --context staging apply -f nginx-secrets.yml
rm nginx-secrets.yml
```

Comment (dleve123) [reply docs/review_app_setup.sh]: Can you elaborate on the problem here? Running this script as is _seemed_ to build a "working" instance of Force, so I'm not aware of a problem that needs to be solved, but I might be missing something.

**Comment by dleve123** [comment]
@eessex @izakp Updated to address most of your comments, but as commented, don't totally understand https://github.com/artsy/force/pull/4412#discussion_r315787578, so didn't address that yet.

File: docs/review_app_setup.sh (hunk 7 above)
Comment (dleve123) [reply docs/review_app_setup.sh]: @izakp Merging this anyways considering that this is working, but happy to iterate from here when you get back on the above.

Comment (izakp) [reply docs/review_app_setup.sh]: Ah okay maybe the certs are not necessary

Comment (izakp) [reply docs/review_app_setup.sh]: Actually yes they are if referenced in the k8s config file - see https://artsy.slack.com/archives/C02BC3HEJ/p1566918604140600?thread_ts=1564512635.190400&cid=C02BC3HEJ
